src/screens/Exchange/Exchange.js

```diff
@@ -232,72 +179,174 @@ class ExchangeScreen extends React.Component<Props, State> {
-    const toAssetSymbol = get(this.state, 'value.toInput.selector.symbol');
-    const prevToAssetSymbol = get(prevState, 'value.toInput.selector.symbol');
+    const {
+      fromAsset, toAsset, fromAmount, errorMessage,
+    } = this.state;
+    const {
+      fromAsset: prevFromAsset, toAsset: prevToAsset, fromAmount: prevFromAmount, errorMessage: prevErrorMessage,
+    } = prevState;
 
     // update from and to options when (supported) assets changes or user selects an option
     if (assets !== prevProps.assets || exchangeSupportedAssets !== prevProps.exchangeSupportedAssets
-      || fromAssetSymbol !== prevFromAssetSymbol || toAssetSymbol !== prevToAssetSymbol) {
-      this.provideOptions();
+      || fromAsset !== prevFromAsset || toAsset !== prevToAsset) {
+      this.options = this.provideOptions();
+    }
+
+    if (!prevProps.hasSeenExchangeIntro && this.props.hasSeenExchangeIntro) {
+      setTimeout(this.focusInputWithKeyboard, 300);
     }
 
-    if (prevProps.oAuthAccessToken !== oAuthAccessToken) {
+    if (
       // access token has changed, init search again
+      (prevProps.oAuthAccessToken !== oAuthAccessToken) ||
+      // valid input provided or asset changed
+      ((
+        fromAsset !== prevFromAsset ||
+        toAsset !== prevToAsset ||
+        fromAmount !== prevFromAmount ||
+        (prevErrorMessage && !errorMessage)) &&
+      validateInput(fromAmount, fromAsset, toAsset, errorMessage))
+    ) {
       this.resetSearch();
       this.triggerSearch();
     }
-
-    if (!prevProps.hasSeenExchangeIntro && this.props.hasSeenExchangeIntro) {
-      setTimeout(this.focusInputWithKeyboard, 300);
-    }
   }
 
-  handleSellMax = () => {
-    const { balances } = this.props;
-    const selectedAssetSymbol = this.getSelectedFromAssetSymbol();
-    const chosenAssetBalance = formatAmount(getBalance(balances, selectedAssetSymbol));
-    const value = { ...this.state.value };
-    value.fromInput.input = chosenAssetBalance;
-    this.handleFormChange(value);
+  getInitialAssets = (): Object => {
+    const { navigation, exchangeSearchRequest } = this.props;
+    const fromAssetCode = navigation.getParam('fromAssetCode') || exchangeSearchRequest?.fromAssetCode || ETH;
+    const toAssetCode = navigation.getParam('toAssetCode') || exchangeSearchRequest?.toAssetCode || PLR;
+    return {
+      fromAsset: this.options.fromOptions.find(a => a.value === fromAssetCode),
+      toAsset: this.options.toOptions.find(a => a.value === toAssetCode),
+    };
   }
 
-  shouldShowSellMax = () => {
-    const { balances } = this.props;
-    const selectedAssetSymbol = this.getSelectedFromAssetSymbol();
-    const assetBalance = getBalance(balances, selectedAssetSymbol);
-    return !!assetBalance;
-  }
+  handleBuySellSwap = () => {
+    const { fromAsset, toAsset } = this.state;
+    this.setState({
+      toAsset: fromAsset,
+      fromAsset: toAsset,
+      fromAmount: '',
```

Comment:
Oh, one more thing I've noticed: 
even though this switch resets fromAmount value, insufficient balance error stays after switching.

Just an idea, maybe errors can be reset in "resetSearch" method?